pkg/ccl/backupccl/restore_span_covering.go

```diff
@@ -307,6 +311,96 @@ func makeEntry(start, end roachpb.Key, f execinfrapb.RestoreFileSpec) execinfrap
 	}
 }
 
+// spanCoveringFilter holds metadata that filters which backups and required spans are used to
+// populate a restoreSpanEntry
+type spanCoveringFilter struct {
+	checkpointFrontier       *spanUtils.Frontier
+	highWaterMark            roachpb.Key
+	introducedSpanFrontier   *spanUtils.Frontier
+	useFrontierCheckpointing bool
+	targetSize               int64
+}
+
+func makeSpanCoveringFilter(
+	checkpointFrontier *spanUtils.Frontier,
+	highWater roachpb.Key,
+	introducedSpanFrontier *spanUtils.Frontier,
+	targetSize int64,
+	useFrontierCheckpointing bool,
+) (spanCoveringFilter, error) {
+	sh := spanCoveringFilter{
+		introducedSpanFrontier:   introducedSpanFrontier,
+		targetSize:               targetSize,
+		highWaterMark:            highWater,
+		useFrontierCheckpointing: useFrontierCheckpointing,
+		checkpointFrontier:       checkpointFrontier,
+	}
+	return sh, nil
+}
+
+// filterCompleted returns the subspans of the requiredSpan that still need to be
+// restored.
+func (f spanCoveringFilter) filterCompleted(requiredSpan roachpb.Span) roachpb.Spans {
+	if f.useFrontierCheckpointing {
+		return f.findToDoSpans(requiredSpan)
+	}
+	if requiredSpan.EndKey.Compare(f.highWaterMark) < 0 {
```

Comment:
huh, yeah it can be `<=0`. In other words, if `requiredSpan.EndKey.Compare(f.highWaterMark) == 0`, the last key in the required span was persisted. This is a small bug in 22.2 and earlier.